src/contrib/cluster/Akka.Cluster.Metrics/Routing/ClusterMetricsRouting.cs

```diff
@@ -0,0 +1,429 @@
+// //-----------------------------------------------------------------------
+// // <copyright file="ClusterMetricsRouting.cs" company="Akka.NET Project">
+// //     Copyright (C) 2009-2019 Lightbend Inc. <http://www.lightbend.com>
+// //     Copyright (C) 2013-2019 .NET Foundation <https://github.com/akkadotnet/akka.net>
+// // </copyright>
+// //-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Akka.Actor;
+using Akka.Annotations;
+using Akka.Cluster.Metrics.Events;
+using Akka.Cluster.Metrics.Serialization;
+using Akka.Cluster.Routing;
+using Akka.Configuration;
+using Akka.Dispatch;
+using Akka.Routing;
+using Akka.Util;
+using Akka.Util.Extensions;
+
+namespace Akka.Cluster.Metrics
+{
+    /// <summary>
+    /// Load balancing of messages to cluster nodes based on cluster metric data.
+    ///
+    /// It uses random selection of routees based on probabilities derived from the remaining capacity of corresponding node.
+    /// </summary>
+    public sealed class AdaptiveLoadBalancingRoutingLogic : RoutingLogic
+    {
+        private readonly ActorSystem _system;
+        private readonly IMetricsSelector _metricsSelector;
+        private Cluster _cluster;
+        private readonly AtomicReference<Tuple<ImmutableArray<Routee>, IImmutableSet<NodeMetrics>, Option<WeightedRoutees>>> _weightedRouteesRef;
+
+        /// <summary>
+        /// Creates instance if <see cref="AdaptiveLoadBalancingRoutingLogic"/>
+        /// </summary>
+        /// <param name="system">The actor system hosting this router</param>
+        /// <param name="metricsSelector">
+        /// Decides what probability to use for selecting a routee, based on remaining capacity as indicated by the node metrics
+        /// </param>
+        public AdaptiveLoadBalancingRoutingLogic(ActorSystem system, IMetricsSelector metricsSelector = null)
+        {
+            _system = system;
+            _metricsSelector = metricsSelector ?? MixMetricsSelector.Instance;
+            _cluster = Cluster.Get(system);
+            
+            // The current weighted routees, if any. Weights are produced by the metricsSelector
+            // via the metricsListener Actor. It's only updated by the actor, but accessed from
+            // the threads of the sender()s.
+            _weightedRouteesRef = new AtomicReference<Tuple<ImmutableArray<Routee>, IImmutableSet<NodeMetrics>, Option<WeightedRoutees>>>(
```

Comment:
I wish I could - but `AtomicReference` requires it's type argument to be `class`, which is not the case for `ValueTuple`. 
For this reason I had to stick with `Tuple` class in some other parts of code base once was migrating it to `ValueTuple`.
P.S. And in turn `AtomicReference` can not use `struct` types since `Interlocked.CompareExchange` requires a reference.